### PR TITLE
TypeScript locale completion

### PR DESCRIPTION
**Pull request by sanchitmalhotra126**
Leveraging TypeScript to get autocompletion and type checking for locale objects. This is done by creating an object type that uses the keys from a given JSON file as its own keys. I've included some examples of using these objects in the Music Lab directory as well. See videos below for examples.

The main open question I have here is if it's safe to import a file outside of `apps/src` in the `apps/src` directory. In this case, we're reading the keys from JSON files in `apps/i18n/...`. Since these files are only referenced in types however, these checks only happen compile-time, and at runtime, importing these locale objects is essentially equivalent to just `import <locale> from '@cdo/locale/<app>'`. 

For example, a TypeScript source file which references `apps/i18n/...` and looks like this

Screenshot 2023-09-05 at 3 47 14 PM

gets converted down to just this at runtime:

Screenshot 2023-09-05 at 3 46 17 PM

Autocompletion/type-checking feature examples in VS Code
- Using common strings:
 
https://github.com/code-dot-org/code-dot-org/assets/85528507/6060e6c5-62c9-4543-a95b-0cd2a27b6d62

- Using app-specific (music) strings:

https://github.com/code-dot-org/code-dot-org/assets/85528507/d13487d6-a835-44af-9e09-82032af7b3f6

### CODE REVIEW

File: apps/src/music/locale.ts [hunk 1]
```diff
@@ -1,8 +1,9 @@
 /**
  * A TypeScript wrapper for the Music Lab locale object which casts
- * it to the {@link MusicLocale} type.
+ * it to the {@link Locale} type.
  */
-import {MusicLocale} from './types';
```
Comment (fisher-alice) [inline apps/src/music/locale.ts:5]: Does this mean you can remove the `MusicLocale` type from `music/types.ts`?

File: apps/src/types/locale.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+/** Type of the locale object, parameterized by a StringMap type (usually derived from JSON) */
+export type Locale<StringMap> = {
+  [key in keyof StringMap]: (replaceMap?: {[key: string]: string}) => string;
+};
+
+/** Exporting the common locale object as the Locale type, with keys derived from the common strings JSON file */
+export const commonI18n = require('@cdo/locale') as Locale<
+  typeof import('@cdo/i18n/common/en_us.json')
+>;
```
Comment (fisher-alice) [inline apps/src/types/locale.ts:24]: So as we convert other .js files to .ts files, we would add other locale objects such as applab locale?

**Review by fisher-alice — COMMENTED** [review]
This is super cool! And thanks for the screencast videos. Just left a couple questions.

Also - when I checked out this branch locally, I didn't see the string value when I hovered but I saw in your screencast that it was included. Unsure if I'm missing a step locally.

Screenshot of `commonI18n.addingToTable()`:

Screen Shot 2023-09-06 at 4 30 09 PM

File: apps/src/music/locale.ts (hunk 1 above)
Comment (sanchitmalhotra126) [reply apps/src/music/locale.ts:5]: Yep we can, good call!

File: apps/src/types/locale.ts (hunk 2 above)
Comment (sanchitmalhotra126) [reply apps/src/types/locale.ts:24]: Yeah, that's what I'm thinking. We can use these objects in JS files as well and should get some intellisense autocompletion, we just won't get any strict type checking from the TS compiler

Comment (fisher-alice) [reply apps/src/types/locale.ts:24]: That would be SO useful - thanks for taking this on!

**Comment by sanchitmalhotra126** [comment]
> This is super cool! And thanks for the screencast videos. Just left a couple questions.
> 
> Also - when I checked out this branch locally, I didn't see the string value when I hovered but I saw in your screencast that it was included. Unsure if I'm missing a step locally.
> 
> Screenshot of `commonI18n.addingToTable()`:
> 
> Screen Shot 2023-09-06 at 4 30 09 PM

Thanks for trying this out! Hmm - is this in a JS or TS file? If it's JS, it's possible that VS Code's intellisense doesn't map all the way back to the JSON file?

**Comment by fisher-alice** [comment]
> > This is super cool! And thanks for the screencast videos. Just left a couple questions.
> > Also - when I checked out this branch locally, I didn't see the string value when I hovered but I saw in your screencast that it was included. Unsure if I'm missing a step locally.
> > Screenshot of `commonI18n.addingToTable()`:
> > Screen Shot 2023-09-06 at 4 30 09 PM
> 
> Thanks for trying this out! Hmm - is this in a JS or TS file? If it's JS, it's possible that VS Code's intellisense doesn't map all the way back to the JSON file?

It was in `SaveStatus.tsx`. But I could click on it and then it took me to the .json file.

**Comment by sanchitmalhotra126** [comment]
> > > This is super cool! And thanks for the screencast videos. Just left a couple questions.
> > > Also - when I checked out this branch locally, I didn't see the string value when I hovered but I saw in your screencast that it was included. Unsure if I'm missing a step locally.
> > > Screenshot of `commonI18n.addingToTable()`:
> > > Screen Shot 2023-09-06 at 4 30 09 PM
> > 
> > 
> > Thanks for trying this out! Hmm - is this in a JS or TS file? If it's JS, it's possible that VS Code's intellisense doesn't map all the way back to the JSON file?
> 
> It was in `SaveStatus.tsx`. But I could click on it and then it took me to the .json file.

Ah gotcha - I was holding down the command key while hovering over it (cmd+click allows me to go to definition) which is when I saw the original string pop up. If I just hover, then I see the same modal. You might need to clean and rebuild, and reopen VS Code as well? Sometimes I need to do that to get changes to show up.

**Review by fisher-alice — APPROVED** [review]
Thanks for answering my questions - this will be super helpful!

File: apps/tsconfig.json [hunk 3]
```diff
@@ -26,6 +26,7 @@
     "paths": {
       "@cdo/apps/*": ["./src/*"],
       "@cdo/static/*": ["./static/*"],
+      "@cdo/i18n/*": ["./i18n/*"],
```
Comment (sanchitmalhotra126) [inline apps/tsconfig.json:29]: @code-dot-org/platform specifically wanted to get sign off on this path alias. To clarify, this only affects code at compile-time and won't actually replace any paths at build-time.

**Review by molly-moen — APPROVED** [review]
Awesome!

**Review by mgc1194 — APPROVED** [review]
LGTM.

**Pull request merged**